subctl/diagnore-pre-install.md

```diff
@@ -0,0 +1,113 @@
+    * "The cluster is deployed with kube-proxy ipvs mode which Submariner does not support"
+* Broker connectivity (New check)
+  * Make sure the broker is reachable from the cluster.
+* K8s Version (Reuse as is)
+  * Query K8s and check we have a minimal supported version (1.17).
+* CNI (Reuse if possible with refactor)
+  * Check if CNI as detected by us is supported (from a list of plugins).
+    * Reuse submariner-operator/pkg/discovery/network -> Discover
+  * Warn that Submariner might not work if the CNI not detected (but don't fail the check).
+  * Extra checks for:
+    * OVN-K:
+      * Version of NB DB is supported.
+    * Calico:
+      * IP Pools used on node labeled as gateway are disabled.
+* Overlapping CIDRs (New check)
+  * Check that the CIDRs aren't overlapping with any known CIDRs from the broker.
+  * If the clusterset is using globalnet, skip this check.
+* Firewall ports (Might be possible to reuse, at least part of it)
+  * Internal VxLAN port:
+    * Skipped on OVN-K
+    * Would be possible to do with:
+      * Spawn on a gateway labeled node a pod running `tcpdump -ln -c 5 -i any udp and src port 9898 and dst port 4500`
+      * Spawn on non-gateway a pod running `for i in $(seq 10); do timeout 2 nc -u -p 9898 <gatewayPodIP> 4500; done`
+  * Public tunnel port:
+    * Skipped when deploying with Load Balancer
+    * Quite problematic to reuse as we don't necessarily have the public IP, we would need to figure out how to deduce it.
+    * Also the existing diagnose check requires a second cluster...
+      * Although we could try to initiate from the same cluster to an external IP, but the success depends on the routing policy.
+    * Spawn on a gateway labeled node on the local cluster a pod running
+      `tcpdump -ln -Q in -A -s 100 -i any udp and dst port <dest port> | grep '<client message>'`
+      where the client message is a random string (part of UUID).
+    * Spawn on a non gateway node on the local cluster running
+      `echo <client message> | for i in $(seq 5); do timeout 2 nc -n -p 9898 -u <gateway ip> <dest port>; done`
+      where the gateway ip is the public IP of the pod running the "sniffer"
+  * NATT port - Same as tunnel port:
+    * Skipped when deploying with Load Balancer
+    * Same exact check as the tunnel port, but with the NATT port number
+
+### Pros
+
+Users will have a tool to check if the infrastructure can support Submariner before actually trying to install it, saving them time and
+headaches trying to debug issues only post-installation.
+
+### Cons
+
+We would have more code to maintain (although sharing with existing diagnose commands could minimize the impact).
+
+### Backward Compatibility
+
+As this is a new command, no impact on existing behavior is expected.
+
+### Alternatives
+
+We could simply focus our efforts on expanding `subctl diagnose` to cover more cases instead, at the cost of forcing users to first install
```

Comment:
This is not a bad idea. If we do not have a fool proof solution to validate the firewall ports, IMO the benefit of this additional command might be somewhat limited because the other commands like K8s version, CNI, Gateway node count etc are mostly well documented.

In fact, we can enhance the `subctl diagnose` to validate the possible checks depending on the state of Submariner deployment (i.e., allow the user to run the same command even when Submariner is not deployed). 
If Submariner is not installed, it will validate whatever is technically feasible based on the available information from the BrokerInfo/user-input. OTOH, if Submariner is installed, it can validate much more than the pre-requisites captured in this EP - like firewall ports, service-discovery objects, state of SM deployment, globalnet objects etc.

Thoughts?

Reply:
It seems there's more support to add all these checks as part of `join`, hence I'm updating the EP to reflect this direction.